fix: fix segmentation fault due to the use of an invalid grpc server_ object.

marete opened the pull request:
BuildAndStart() can fail for any number of reasons (e.g. when passed a port reserved for privileged use or a non-existent local address). We were not checking for this and would try to use the server unconditionally, causing a segmentation fault (e.g. when emulator was called by a non-root user as: ./emulator -p 100).

This adds a fix so that we can check that we have a valid server object before trying to use it.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/24)
<!-- Reviewable:end -->

marete committed:
fix: fix segmentation fault due to the use of an invalid grpc server_ object.

BuildAndStart() can fail for any number of reasons (e.g. when passed a
port reserved for privileged use or a non-existent local address). We
were not checking for this and would try to use the server
unconditionally, causing a segmentation fault (e.g. when emulator was
called by a non-root user as: ./emulator -p 100).

This adds a fix so that we can check that we have a valid server
object before trying to use it.

prawilny reviewed (commented):
In general it looks good, but I left some comments.
You are free to merge after addressing them (nitpicks are ignorable).

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/24#-:-OU_XfubBI_9nXtEY769:b8sbqi6)* status: all files reviewed, 4 unresolved discussions (waiting on @dopiera)

___
*[`google/cloud/bigtable/emulator/emulator.cc` line 40 at r1](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/24#-OU_WKB83ZDJU4hgLWZ_:-OU_WKB97c8VBPcC65a_:b-wn9gh9) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8aeae583bb29efa7df0f8e89d8774bb7eb345e4e/google/cloud/bigtable/emulator/emulator.cc#L40)):*
> ```cpp
>           absl::GetFlag(FLAGS_host), absl::GetFlag(FLAGS_port));
>   if (!maybe_server) {
>     std::cerr << "CreateDefaultEmulatorServer() failed. See other logging for "
> ```

nit: "other logging" -> "the logs"?
___
*[`google/cloud/bigtable/emulator/emulator.cc` line 43 at r1](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/24#-OU_WNjc5ty1-spIHwwa:-OU_WNjc5ty1-spIHwwb:bwjezau) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8aeae583bb29efa7df0f8e89d8774bb7eb345e4e/google/cloud/bigtable/emulator/emulator.cc#L43)):*
> ```cpp
>                  "possible reason"
>               << std::endl;
>     std::exit(1);
> ```

Why not just `return 1`?
___
*[`google/cloud/bigtable/emulator/server.cc` line 336 at r1](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/24#-OU_XJu-Ex7YyNLoweyE:-OU_XJu0DUrhVcQYu9YR:bx561gf) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8aeae583bb29efa7df0f8e89d8774bb7eb345e4e/google/cloud/bigtable/emulator/server.cc#L336)):*
> ```cpp
>   void Shutdown() override { server_->Shutdown(); }
>   void Wait() override { server_->Wait(); }
>   bool HasValidServer() { return static_cast<bool>(server_); }
> ```

nit: I feel like this name is strange, but have no idea how to name it better.
___
*[`google/cloud/bigtable/emulator/server_test.cc` line 35 at r1](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/24#-OU_WoD63tsFJO7svvGk:-OU_WoD63tsFJO7svvGl:bfl0vj2) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8aeae583bb29efa7df0f8e89d8774bb7eb345e4e/google/cloud/bigtable/emulator/server_test.cc#L35)):*
> ```cpp
>   std::shared_ptr<grpc::Channel> channel_;
>   grpc::ClientContext ctx_;
> ```

Maybe let's add a test that ensures that the error branch doesn't crash?
We could just start the emulator on a single port twice.


<!-- Sent from Reviewable.io -->

marete committed:
fix: emulator: server: various fixes for review.

Most importantly, we implement a test to ensure that we don't crash
when GRPC server creation fails.

Other fixes implementing review comments.

Signed-off-by: Brian Gitonga Marete <bgmarete@gmail.com>

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/24)* status: 2 of 4 files reviewed, 2 unresolved discussions (waiting on @dopiera and @prawilny)

___
*[`google/cloud/bigtable/emulator/emulator.cc` line 43 at r1](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/24#-OU_WNjc5ty1-spIHwwa:-OU_ZrwD4mlXHJOnXxiD:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8aeae583bb29efa7df0f8e89d8774bb7eb345e4e/google/cloud/bigtable/emulator/emulator.cc#L43)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

Why not just `return 1`?
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/server_test.cc` line 35 at r1](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/24#-OU_WoD63tsFJO7svvGk:-OU_e96D3WIzxrDLaW9v:b7vtxb8) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8aeae583bb29efa7df0f8e89d8774bb7eb345e4e/google/cloud/bigtable/emulator/server_test.cc#L35)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

Maybe let's add a test that ensures that the error branch doesn't crash?
We could just start the emulator on a single port twice.
</blockquote></details>

Used an invalid host interface address for the test of the failed server creation path instead.

Note that the ports are bound to using `SO_REUSEADDR`, so starting another server on the same port will very happily succeed. 


<!-- Sent from Reviewable.io -->
